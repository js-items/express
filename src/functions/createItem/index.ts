import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { CREATED } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import sendResponse from '../../utils/sendResponse';

const createItem: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeCreateItem
  );

  await transactionHandler({ req, res }, async () => {
    const { item } = await config.service.createItem({
      id: req.body.id,
      item: config.convertDocumentIntoItem({ document: req.body, req, res }),
    });

    sendResponse({
      config,
      req,
      res,
      responseObject: config.convertItemIntoDocument({ item, req, res }),
      status: CREATED,
    });
  });
};

export default createItem;
