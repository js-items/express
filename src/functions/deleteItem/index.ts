import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { NO_CONTENT } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const deleteItem: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeDeleteItem
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');
    
    await config.service.deleteItem({
      filter: config.createFilter({ filter, req, res }),
      id: req.params.id,
    });

    sendResponse({
      config,
      req,
      res,
      status: NO_CONTENT,
    });
  });
};

export default deleteItem;
