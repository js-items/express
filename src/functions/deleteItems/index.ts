import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { NO_CONTENT } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const deleteItems: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeDeleteItems
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');
    
    await config.service.deleteItems({
      filter: config.createFilter({ filter, req, res }),
    });

    sendResponse({
      config,
      req,
      res,
      status: NO_CONTENT,
    });
  });
};

export default deleteItems;
