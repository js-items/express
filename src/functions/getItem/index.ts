import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { OK } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const getItem: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeGetItem
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');

    const { item } = await config.service.getItem({
      filter: config.createFilter({ filter, req, res }),
      id: req.params.id,
    });
    
    sendResponse({
      config,
      req,
      res,
      responseObject: config.convertItemIntoDocument({ item, req, res }),
      status: OK,
    });
  });
};

export default getItem;
