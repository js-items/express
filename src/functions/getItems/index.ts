import { Item } from '@js-items/foundation';
import boolean from 'boolean';
import { Request, Response } from 'express';
import { OutgoingHttpHeaders } from 'http';
import { OK } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import _isNil from 'ramda/src/isNil';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import getNumberQueryParam from '../../utils/getNumberQueryParam';
import sendResponse from '../../utils/sendResponse';

const getItems: RequestHandlerFactory = <I extends Item>(
  config: FacadeConfig<I>
) => async (req: Request, res: Response) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(
    config.beforeGetItems
  );

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');

    const sort = !_isNil(req.query.sort)
      ? getJsonQueryParam(req.query, 'sort')
      : config.defaultSort;
      
    const limit = getNumberQueryParam(
      req.query,
      'limit',
      config.defaultPaginationLimit
    );
    
    const createdFilter = config.createFilter({ filter, req, res });

    const { cursor, items } = await config.service.getItems({
      filter: createdFilter,
      pagination: {
        after: req.query.after,
        before: req.query.before,
        limit,
      },
      sort,
    });

    const { count: totalCount } = await config.service.countItems({
      filter: createdFilter,
    });

    const responseHeaders: OutgoingHttpHeaders = {
      [config.afterHeaderName]: cursor.after,
      [config.beforeHeaderName]: cursor.before,
      [config.hasBeforeHeaderName]: cursor.hasBefore.toString(),
      [config.hasAfterHeaderName]: cursor.hasAfter.toString(),
      [config.totalHeaderName]: totalCount,
    };

    const responseData = items.map(item =>
      config.convertItemIntoDocument({ item, req, res })
    );

    const nestedObject = {
      [config.paginationKey]: {
        [config.afterKey]: cursor.after,
        [config.beforeKey]: cursor.before,
        [config.hasBeforeKey]: cursor.hasBefore,
        [config.hasAfterKey]: cursor.hasAfter,
        [config.totalKey]: totalCount,
      },
      [config.dataKeyName]: responseData,
    };

    const enveloped = boolean(req.query[config.envelopeParamName]);

    const responseObject = enveloped ? nestedObject : responseData;
    const headers = enveloped ? {} : responseHeaders;

    sendResponse({ req, res, config, status: OK, headers, responseObject });
  });
};

export default getItems;
