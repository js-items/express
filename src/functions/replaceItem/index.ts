import { Item } from '@js-items/foundation';
import { Request, Response } from 'express';
import { OK } from 'http-status-codes';
import _defaultTo from 'ramda/src/defaultTo';
import FacadeConfig from '../../FacadeConfig';
import RequestHandlerFactory from '../../types/RequestHandlerFactory';
import getJsonQueryParam from '../../utils/getJsonQueryParam';
import sendResponse from '../../utils/sendResponse';

const replaceItem: RequestHandlerFactory =  <I extends Item>(config: FacadeConfig<I>) => async (
  req: Request,
  res: Response
) => {
  const transactionHandler = _defaultTo(config.defaultTransactionHandler)(config.beforeReplaceItem);

  await transactionHandler({ req, res }, async () => {
    const filter = getJsonQueryParam(req.query, 'filter');

    const result = await config.service.replaceItem({
      filter: config.createFilter({ filter, req, res }),
      id: req.params.id,
      item: config.convertDocumentIntoItem({ document: req.body, req, res }),
    });

    sendResponse({
      config,
      req,
      res,
      responseObject: config.convertItemIntoDocument({ item: result.item, req, res }),
      status: OK,
    });
  });
};

export default replaceItem;