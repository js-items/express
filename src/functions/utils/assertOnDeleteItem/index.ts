// tslint:disable:no-any
import { DeleteItem } from '@js-items/foundation';
import { TestItem } from '@js-items/foundation/dist/functions/utils/testItem';
import { NO_CONTENT } from 'http-status-codes';
import supertest from 'supertest';
import createTestUrl from '../createTestUrl';

const assertOnDeleteItem = async <I extends TestItem>({
  client,
  status = NO_CONTENT,
  params,
  expectedParams,
  url,
  deleteItem,
}: Options<I>) => {
  const requestUrl =
    params !== undefined ? createTestUrl({ url, params }) : url;

  const result = await client.delete(requestUrl).send();

  expect(result.status).toBe(status);
  expect(result.body).toMatchSnapshot();

  if (expectedParams !== undefined) {
    expect(deleteItem).toBeCalledWith(expectedParams);
  }
};

export interface Options<I extends TestItem> {
  readonly client: supertest.SuperTest<supertest.Test>;
  readonly status?: number;
  readonly params?: any;
  readonly expectedParams?: { filter: any; pagination: any; sort: any };
  readonly url: string;
  readonly deleteItem: DeleteItem<I>;
}

export default assertOnDeleteItem;
