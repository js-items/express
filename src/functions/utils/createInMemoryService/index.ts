// tslint:disable:no-object-literal-type-assertion
import { Item } from '@js-items/foundation';
import Facade from '@js-items/foundation/dist/Facade';
import {
  Options as CreateItemOptions,
  Result as CreateItemResult,
} from '@js-items/foundation/dist/functions/CreateItem';
import {
  Options as GetItemOptions,
  Result as GetItemResult,
} from '@js-items/foundation/dist/functions/GetItem';
import { Result as GetItemsResult } from '@js-items/foundation/dist/functions/GetItems';
import { Result as ReplaceItemResult } from '@js-items/foundation/dist/functions/ReplaceItem';
import { Result as UpdateItemResult } from '@js-items/foundation/dist/functions/UpdateItem';
import { start } from '@js-items/foundation/dist/interfaces/Cursor';
import FactoryConfig from './FactoryConfig';

const defaultPromise = () => Promise.resolve();
/* istanbul ignore next */
const createInMemoryService = <I extends Item>(
  factoryConfig: FactoryConfig<I>
): Facade<I> => ({
  countItems: () => Promise.resolve({ count: 1 }),
  createItem: ({ item }: CreateItemOptions<I>) =>
    Promise.resolve({ item } as CreateItemResult<I>),
  deleteItem: defaultPromise,
  deleteItems: defaultPromise,
  getItem: ({ id }: GetItemOptions<I>) =>
    Promise.resolve({ item: { id } } as GetItemResult<I>),
  getItems: () =>
    Promise.resolve({
      cursor: {
        after: start,
        before: start,
        hasAfter: false,
        hasBefore: false,
      },
      items: [{ id: '1' }, { id: '2' }],
    } as GetItemsResult<I>),
  replaceItem: () =>
    Promise.resolve({ item: { id: '2' } } as ReplaceItemResult<I>),
  updateItem: () =>
    Promise.resolve({ item: { id: '1' } } as UpdateItemResult<I>),
  ...factoryConfig,
});

export default createInMemoryService;
