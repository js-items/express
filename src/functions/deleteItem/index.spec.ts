// tslint:disable:no-any
import * as sourceMapSupport from 'source-map-support';
import assertOnDeleteItem from '../utils/assertOnDeleteItem';
import initTests from '../utils/initTests';
sourceMapSupport.install();
import dotenv from 'dotenv';
dotenv.config();
import { ItemNotFoundError } from '@js-items/foundation';
import { Options as DeleteItemOptions } from '@js-items/foundation/dist/functions/DeleteItem';
import testItem, {
  TestItem,
} from '@js-items/foundation/dist/functions/utils/testItem';
import { NOT_FOUND, OK } from 'http-status-codes';
import { TEST_URL } from '../../constants';
import createInMemoryService from '../utils/createInMemoryService';

jest.mock('uuid', () => ({
  v4: jest.fn(() => '1'),
}));

const notExistingId = 'not-existing-id';

const deleteItem = jest.fn(({ filter, id }: DeleteItemOptions<TestItem>) => {
  if (
    id !== testItem.id ||
    (filter !== undefined &&
      (filter as any).booleanProperty !== testItem.booleanProperty)
  ) {
    return Promise.reject(new ItemNotFoundError('item', testItem.id));
  }

  return Promise.resolve();
});

describe('@deleteItem', () => {
  beforeEach(async () => {
    jest.clearAllMocks();
  });

  const service = createInMemoryService({
    deleteItem,
  });

  const { request } = initTests<TestItem>({ service });

  const defaultParams = {
    filter: JSON.stringify({ booleanProperty: testItem.booleanProperty }),
  };

  const defaultOptions = {
    client: request,
    deleteItem,
    params: defaultParams,
    url: `${TEST_URL}/${testItem.id}`,
  };

  it('delete item', async () => {
    await assertOnDeleteItem(defaultOptions);
  });

  it('throws not found error when item does not exist', async () => {
    await assertOnDeleteItem({
      ...defaultOptions,
      status: NOT_FOUND,
      url: `${TEST_URL}/${notExistingId}`,
    });
  });

  it('throws not found error when item does not exist because filter does not match', async () => {
    await assertOnDeleteItem({
      ...defaultOptions,
      params: {
        ...defaultOptions.params,
        filter: JSON.stringify({
          booleanProperty: !testItem.booleanProperty,
        }),
      },
      status: NOT_FOUND,
      url: `${TEST_URL}/${testItem.id}`,
    });
  });

  it('throws not found error when item does not exist and envelope is enabled and pretty is disabled', async () => {
    await assertOnDeleteItem({
      ...defaultOptions,
      params: {
        ...defaultOptions.params,
        envelope: true,
        pretty: false,
      },
      status: OK,
      url: `${TEST_URL}/${notExistingId}`,
    });
  });

  it('delete item when envelope is enabled and pretty response is enabled', async () => {
    await assertOnDeleteItem({
      ...defaultOptions,
      params: {
        ...defaultParams,
        envelope: true,
      },
      status: OK,
    });
  });

  it('delete item when envelope is enabled and pretty response is disabled', async () => {
    await assertOnDeleteItem({
      ...defaultOptions,
      params: {
        ...defaultParams,
        envelope: true,
        pretty: false,
      },
      status: OK,
    });
  });
// tslint:disable-next-line:max-file-line-count
});
