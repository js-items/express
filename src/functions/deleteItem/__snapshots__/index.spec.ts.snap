// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@deleteItem delete item 1`] = `Object {}`;

exports[`@deleteItem delete item when envelope is enabled and pretty response is disabled 1`] = `
Object {
  "headers": Object {
    "x-powered-by": "Express",
  },
  "response": Object {},
  "status": 204,
}
`;

exports[`@deleteItem delete item when envelope is enabled and pretty response is enabled 1`] = `
Object {
  "headers": Object {
    "x-powered-by": "Express",
  },
  "response": Object {},
  "status": 204,
}
`;

exports[`@deleteItem throws not found error when item does not exist 1`] = `
Object {
  "itemId": "6da9dce6-be40-4fac-bd9c-6a2f348a1c87",
  "itemName": "item",
  "transactionId": "1",
}
`;

exports[`@deleteItem throws not found error when item does not exist and envelope is enabled and pretty is disabled 1`] = `
Object {
  "headers": Object {
    "x-powered-by": "Express",
  },
  "response": Object {
    "itemId": "6da9dce6-be40-4fac-bd9c-6a2f348a1c87",
    "itemName": "item",
    "transactionId": "1",
  },
  "status": 404,
}
`;

exports[`@deleteItem throws not found error when item does not exist because filter does not match 1`] = `
Object {
  "itemId": "6da9dce6-be40-4fac-bd9c-6a2f348a1c87",
  "itemName": "item",
  "transactionId": "1",
}
`;
